File: Bronze1-1/06-21-pm/unit04/consecutive-subarrays/consecutive.java
/*
 * Remove consecutive subarrays
 */
import java.util.*;
public class consecutive {
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int N = in.nextInt(); //# of integers
		int K = in.nextInt();
		int[] A = new int[N+1];
		for(int j=0; j<N; j++)
			A[j] = in.nextInt();
		A[N] = Integer.MAX_VALUE;

		ArrayList<Integer> B = new ArrayList<>();
		for(int j=0; j<=N; j++) {
			if(B.size()<K || A[j]==1+B.get(B.size()-1)) {
				B.add(A[j]);
			}
			else { //A[j] is not consecutive to B[last]
				//check the consecutive subarray seen so far
				int i = B.size()-2;
				while(i>=0) {
					if(B.get(i)+1!=B.get(i+1))
						break;
					i--;
				}
				if(i<B.size()-K) { 
					//a consecutive subarray with length >= K exists; remove it
					for(int x=B.size()-1; x>i; x--)
						B.remove(B.size()-1);
				}

				B.add(A[j]);
			}
		}

		if(B.get(B.size()-1)==Integer.MAX_VALUE) B.remove(B.size()-1);
		
		if(B.isEmpty())
			System.out.println("NONE");
		else {
			for(int v : B)
				System.out.print(v + " ");
			System.out.println();
		}
		in.close();
	}
}

File: Bronze1-1/06-21-pm/unit04/consecutive-subarrays/consecutive.cpp
/*
 * Remove consecutive subarrays
 */
#include <bits/stdc++.h>
using namespace std;

int A[100005];
int main() {
	int N; cin >> N; //# of integers
	int K; cin >> K;
	for(int j=0; j<N; j++)
		cin >> A[j];
	A[N] = INT_MAX;

	vector<int> B;
	for(int j=0; j<=N; j++) {
		if(B.size()<K || A[j]==1+B.back()) {
			B.push_back(A[j]);
		}
		else { //A[j] is not consecutive to B[last]
			
			//check the consecutive subarray seen so far
			int i = B.size()-2;
			while(i>=0) {
				if(B[i]+1!=B[i+1])
					break;
				i--;
			}
			if(i<B.size()-K) { 
				//a consecutive subarray with length >= K exists; remove it
				B.erase(B.begin()+i+1, B.end());
			}

			B.push_back(A[j]);
		}
	}

	if(B.back()==INT_MAX) B.pop_back();
	
	if(B.empty())
		cout << "NONE\n";
	else {
		for(int v : B)
			cout << v << " ";
		cout << '\n';
	}
}

